dbt_coverage/__init__.py

```diff
@@ -15,82 +15,129 @@
+        return doc is not None and doc != ''
+
+    @staticmethod
+    def is_valid_test(tests):
+        return tests is not None and tests
 
 
 @dataclass
-class Table:
+class Table(EntityWithCoverage):
     """Dataclass containing the information about a database table and its columns."""
 
     name: str
     columns: Dict[str, Column]
 
     @staticmethod
-    def from_catalog_node(node):
+    def from_node(node) -> Table:
         return Table(
             node['metadata']['name'],
             {col['name']: Column.from_node(col) for col in node['columns'].values()}
         )
 
-    @staticmethod
-    def from_manifest_node(node):
-        return Table(
-            node['name'],
-            {col['name']: Column.from_node(col) for col in node['columns'].values()}
-        )
-
     def get_column(self, column_name):
         return self.columns.get(column_name)
 
-    def get_coverage(self):
-        documented = {col.name for col in self.columns.values() if col.has_doc()}
-        total = {col.name for col in self.columns.values()}
-        return documented, total
+    def coverage(self, cov_type: CoverageType):
+        coverages = [col.coverage(cov_type) for col in self.columns.values()]
+        covered = set(f'{self.name}.{col}' for covered, _ in coverages for col in covered)
+        total = set(f'{self.name}.{col}' for _, total in coverages for col in total)
+        return covered, total
 
 
 @dataclass
-class Catalog:
+class Catalog(EntityWithCoverage):
     """Dataclass containing the information about a database catalog, its tables and columns."""
 
     tables: Dict[str, Table]
 
     @staticmethod
-    def from_catalog_nodes(nodes):
-        return Catalog({f"{node['metadata']['name']}": Table.from_catalog_node(node)
-                        for node in nodes})
-
-    @staticmethod
-    def from_manifest_nodes(nodes):
-        return Catalog({f"{node['name']}": Table.from_manifest_node(node)
+    def from_nodes(nodes):
+        return Catalog({f"{node['metadata']['name']}": Table.from_node(node)
                         for node in nodes})
 
     def get_table(self, table_name):
         return self.tables.get(table_name)
 
-    def get_coverage(self):
-        coverages = [(table, table.get_coverage()) for table in self.tables.values()]
-        documented = set(f'{table.name}.{col}'
-                         for table, (documented, _) in coverages
-                         for col in documented)
-        total = set(f'{table.name}.{col}'
-                    for table, (_, total) in coverages
-                    for col in total)
-        return documented, total
+    def coverage(self, cov_type: CoverageType):
+        coverages = [table.coverage(cov_type) for table in self.tables.values()]
+        covered = set(col for covered, _ in coverages for col in covered)
+        total = set(col for _, total in coverages for col in total)
+        return covered, total
+
+
+@dataclass
+class Manifest:
```

Comment:
@sweco You did this very nicely in here but I cannot help but wonder whether `dbt` doesn't do something like this already. Do you think it might make sense to utilize that or would it be too much trouble for potentially nothing?

Reply:
I'm sorry I haven't mentioned this, I really wanted to, but forgot... It's actually all copied from here: https://github.com/dbt-labs/dbt-docs/blob/da84e95ab9febde3127ed28ff698516b2a610532/src/app/services/project_service.js#L149-L215

Reply:
@sweco I see.

I was wondering whether this could just reuse something like https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/manifest.py, so that it's not being re-implemented with each and every change of the manifest 🙁 

But I guess it's not that much of a problem here and generally not worth the trouble :)

Reply:
That's a good idea but I did not find an easy way to load the `manifest.json` into the `Manifest` class so I guess it's easier to keep it as is...

Reply:
Understood -- thanks @sweco, feel free to consider this resolved :)